components/support/sql/src/open_database.rs

```diff
@@ -0,0 +1,544 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/// Use this module to open a new SQLite database connection.
+///
+/// The code handles some common cases:
+///
+///   - Opening new databases.  If this is the first time opening the database, then initialize it
+///     to the current schema.
+///
+///   - Migrating existing databases.  If this is an existing database, then run a series of
+///     upgrade functions to migrate it to the current schema.
+///
+///   - Handling migration failures.   If opening or migrating a database results in an error,
+///     we can optionally delete the database file and create a new one.
```

Comment:
It would be great to have a high-level sketch of how to use the component as part of the module-level doc here. We haven't been great at this in the past, but it's worth us trying to keep getting better over time.

I'm thinking specifically about something narrative to set expectations, along the lines of:

* "You'll need to create a `MigrationLogic` struct, that handles A, B, C."
* "Then you pass this to `open_database` which does X, Y, Z."